sweyer_plugin/android/src/main/kotlin/com/nt4f04und/sweyer/sweyer_plugin/handlers/FetchHandler.kt

```diff
@@ -0,0 +1,233 @@
+            // * DOCUMENT_ID
+            // * HEIGHT
+            // * WIDTH
+            // * INSTANCE_ID
+            //
+            // * IS_ALARM
+            // * IS_AUDIOBOOK
+            // * IS_MUSIC - we fetch only music, see `selection` above
+            // * IS_NOTIFICATION
+            // * IS_PODCAST
+            // * IS_RECORDING
+            // * IS_RINGTONE
+            //
+            // * IS_DOWNLOAD
+            // * IS_DRM
+            //
+            // * IS_TRASHED - trashed items are excluded, see `selection` above
+            // * IS_PENDING - pending items are excluded, see `selection` above
+            //
+            // * MIME_TYPE
+            // * NUM_TRACKS - the number of songs in the origin this media comes from
+            // * ORIENTATION
+            // * ORIGINAL_DOCUMENT_ID
+            // * OWNER_PACKAGE_NAME
+            // * RELATIVE_PATH
+            // * RESOLUTION
+            // * VOLUME_NAME
+            // * XMP
+            // * TITLE_RESOURCE_URI
+            //
+            // * BOOKMARK - position within the audio item at which
+            // playback should be resumed. For me it's making no sense to remember position for each
+            // media item.
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                put(MediaStore.Audio.Media.IS_FAVORITE,
+                    "isFavoriteInMediaStore" to { cursor, index -> cursor.getInt(index) == 1 })
+                put(MediaStore.Audio.Media.GENERATION_ADDED,
+                    "generationAdded" to { cursor, index -> cursor.getInt(index) })
+                put(MediaStore.Audio.Media.GENERATION_MODIFIED,
+                    "generationModified" to { cursor, index -> cursor.getInt(index) })
+                put(MediaStore.Audio.Media.GENRE, "genre" to { cursor, index -> cursor.getString(index) })
+                put(MediaStore.Audio.Media.GENRE_ID, "genreId" to { cursor, index -> cursor.getInt(index) })
+            }
+        })
+    }
+
+    fun retrieveAlbums(resolver: ContentResolver): ArrayList<MutableMap<String, Any?>> {
+        return executeQuery(resolver,
+            MediaStore.Audio.Albums.EXTERNAL_CONTENT_URI,
+            MediaStore.Audio.Albums.ALBUM + " IS NOT NULL",
+            buildMap {
+                put(MediaStore.Audio.Albums._ID, "id" to { cursor, index -> cursor.getInt(index) })
+                put(MediaStore.Audio.Albums.ALBUM, "album" to { cursor, index -> cursor.getString(index) })
+                put(MediaStore.Audio.Albums.ALBUM_ART, "albumArt" to { cursor, index -> cursor.getString(index) })
+                put(MediaStore.Audio.Albums.ARTIST, "artist" to { cursor, index -> cursor.getString(index) })
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                    put(MediaStore.Audio.Albums.ARTIST_ID, "artistId" to { cursor, index -> cursor.getInt(index) })
+                }
```

Comment:
I don't believe this was added in 29, this annotation is likely just incorrect.
I also remember asking this on Reddit and people agreed with me on this.
Did you confirm that on API less than 28 we don't have this column?

Reply:
Actually, we don't even need an additional confirmation.
I have an API 28 Redmi Note 5 and it works just fine fetching artists from albums.
The screenshots from README were made on this device.

I advocate for adding an ignore comment with explanation for this line of code.

Reply:
The Annotation is right, this constant was added in [this commit](https://android.googlesource.com/platform/frameworks/base/+/56c021ffd8490a7cfc815ee76f7109e8a1621df7) in 2019, which is between API 28 and 29. So if we would lour our target sdk level, the code would not compile.

Whether or not the column was available in versions before 29 is of course an entire different question. The problem is that we can't rely on it at all. Fetching a non-existent column will cause an exception:

<details>
<summary>Stacktrace</summary>

```
22:43:56.555 flutter                  I  The following exception was thrown in re-fetch album:
22:43:56.556 flutter                  I  SweyerPluginException: retrieveAlbums failed
                                         
                                         Caused by:
                                         PlatformException(UNEXPECTED_ERROR, Invalid column definetly_not_in_there, java.lang.IllegalArgumentException: Invalid column definetly_not_in_there
                                         	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:172)
                                         	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:142)
                                         	at android.content.ContentProviderProxy.query(ContentProviderNative.java:481)
                                         	at android.content.ContentResolver.query(ContentResolver.java:1219)
                                         	at android.content.ContentResolver.query(ContentResolver.java:1151)
                                         	at android.content.ContentResolver.query(ContentResolver.java:1107)
                                         	at com.nt4f04und.sweyer.sweyer_plugin.handlers.FetchHandler.executeQuery(FetchHandler.kt:221)
                                         	at com.nt4f04und.sweyer.sweyer_plugin.handlers.FetchHandler.retrieveAlbums(FetchHandler.kt:113)
                                         	at com.nt4f04und.sweyer.sweyer_plugin.SweyerPlugin.onMethodCall$lambda-12(SweyerPlugin.kt:176)
                                         	at com.nt4f04und.sweyer.sweyer_plugin.SweyerPlugin.$r8$lambda$TG7SbxIRc7nzpR
22:43:56.556 flutter                  I  #0      MethodChannelSweyerPlugin.retrieveAlbums (package:sweyer_plugin/sweyer_plugin_method_channel.dart:86:7)
                                         <asynchronous suspension>
                                         #1      SweyerPlugin.retrieveAlbums (package:sweyer_plugin/sweyer_plugin.dart:79:18)
                                         <asynchronous suspension>
                                         #2      ContentControl.refetch (package:sweyer/logic/player/content.dart:398:26)
                                         <asynchronous suspension>
                                         #3      Future.wait.<anonymous closure> (dart:async/future.dart:522:21)
                                         <asynchronous suspension>
                                         #4      Future.any.onValue (dart:async/future.dart:612:5)
                                         <asynchronous suspension>
22:43:56.556 flutter                  I  ----------------------------------------------------
```
</details>

I did test it on an API 21 emulator (our current min sdk) and the `artist_id` column was there (the refresh still failed because the first and last year were `null`, but I beliefe that is a separate issue and just related the the audio file I used to test on the emulator).

So should we try fetching albums with the artist id constant regardless of API level, catch the error and in case of failure try again without the artist id?

Reply:
The commit message says "As pointed out by developers, we already have ARTIST, so we should
also have ARTIST_ID.".

This implies that the column existed from the same level as ARTIST, which is API level 1.

Whether this is true or not, I wouldn't be bothered with it, since you verified this is not crashing on API 21, which is the minimum version we support.

I don't think we need to do anything else here, until proven that there's anything broken about it.

Reply:
> This implies that the column existed from the same level as ARTIST, which is API level 1.

I don't think it implies that at all, ARTIST is a completely different column.

> since you verified this is not crashing on API 21

I verified it on **one** emulator, which is not event representative of what is actually out there. Because this column is not part of the API pre 29 there is no guarantee and no obligations for phone vendors to have that column.

I think the catch and retry is giving us the best of both worlds here.

Reply:
Media ContentProvider is backed by a single main [table](https://cs.android.com/android/platform/superproject/+/master:packages/providers/MediaProvider/src/com/android/providers/media/DatabaseHelper.java;l=1014;drc=94833d1e27fad8d86887f70eac7b706e80a69096) which has views into various content types, one of them is [album](https://cs.android.com/android/platform/superproject/+/master:packages/providers/MediaProvider/src/com/android/providers/media/DatabaseHelper.java;l=1485;drc=94833d1e27fad8d86887f70eac7b706e80a69096).

It was implemented like this from API level 1 https://github.com/aosp-mirror/platform_packages_providers_mediaprovider/blame/df8751b0a94d9648c8120342a31782ae2835fba0/src/com/android/providers/media/MediaProvider.java#L929

The constant that was added in 29 is just a missing string in MediaStore namespace, which on its own is merely a set of constants representative of the database implementation. 

> I verified it on **one** emulator, which is not representative of what is actually out there. Because this column is not part of the API pre 29 there is no guarantee and no obligations for phone vendors to have that column.

I don't understand why would this matter.

> I think the catch and retry is giving us the best of both worlds here.

I think we don't need it, since it's impossible for this column to not exist.

Hence I do not agree that the artist can ever be null.

Reply:
Sorry for dragging this out, I just want to explain my thinking:

> The whole MediaStore backend is in AOSP and the API was there from the very beginning.

My problem is that the the specific columns are an implementation detail and not part of the **API** unless the corresponding constant exists. My point is that just because the column existed since the very beginning does not mean it is part of the public facing Android API, relying on it's presence means relying on an implementation detail. 

> If vendors modified this specific Database view and removed this column (I cannot imagine why would they do this), I would consider this an OS bug

This is where we disagree, and I think Google does too. The official Compatibility Test Suite only tests for the existence of this column since [this commit from 2019](https://android.googlesource.com/platform/cts/+/5c1def314db81aef9a3643259d756a4bba6ed349). The point is that if something is not part of the public API, it is an implementation detail and OEMs are free to drop/change it. I also don't see why they should drop it, but relying on an implementation detail is wrong nevertheless and will cause problems at some point.

> I think we don't need it, since it's impossible for this column to not exist.

Even if you think so, could we please keep it? In the normal case this will only cause a few extra CPU cycles for the try catch, I don't think this will have any major impact on CPU usage or battery life.

Reply:
Okay, let's keep it :-)